File: iterator/myiterator.cpp
#include <iostream>
#include <vector>

using namespace std;

class Dish {
    private:
        string name;
    public:
        Dish(string name) : name(name){}
        string getName() {
            return name;
        }
};

class Eating {
    public:
    virtual bool hasNextDish() = 0;
    virtual Dish *nextDish() = 0;

    virtual ~Eating() {}
};

class RestaurantEating : public Eating {
    private:
    vector<Dish> &dishes;
    int index;

    public:
    RestaurantEating(vector<Dish> &dishes) : dishes(dishes),index(0) {}

    bool hasNextDish() override {
        return index < dishes.size();
    }

    Dish *nextDish() override {
        if (hasNextDish()) {
            return &dishes[index++];
        }
        return nullptr;
    }
};

class Dinner {
    public:
    virtual Eating *createDinner() = 0;
    virtual ~Dinner() = default;
};

class WeekendDinner : public Dinner {
    private:
    vector<Dish> items;

    public:
    WeekendDinner(vector<Dish> dishes) : items(dishes) {}

    
    Eating *createDinner() override {
        return new RestaurantEating(items);
    }
};

int main() {
    vector<Dish> dishes;
    dishes.push_back(Dish("Soup"));
    dishes.push_back(Dish("Fries"));
    dishes.push_back(Dish("Sandwich"));
    dishes.push_back(Dish("Dessert"));

    WeekendDinner weekenddinner(dishes);

    Eating *eating = weekenddinner.createDinner();

    while(eating->hasNextDish()) {
        cout << eating->nextDish()->getName() << endl;
    }
    delete eating;

    return (0);
}

File: iterator/swift/myiterator.swift
import Foundation

// 1. Element Protocol
protocol Dish {
    var name: String { get }
}

struct FoodDish: Dish {
    let name: String
}

// 2. Iterator Protocol (Eating)
protocol Eating {
    func eat(dish: Dish)
}

// 3. Concrete Iterator (RestaurantEating)
class RestaurantEating: Eating {
    func eat(dish: Dish) {
        print("Enjoying \(dish.name) at the restaurant.")
    }
}

// 4. Abstract Factory/Creator (Dinner)
protocol Dinner {
    func createDinner() -> Eating
}

// 5. Concrete Collection (WeekendDinner)
class WeekendDinner: Dinner {
    private var dishes: [Dish] = []
    
    func add(dish: Dish) {
        dishes.append(dish)
    }
    
    func createDinner() -> Eating {
        return RestaurantEating()
    }
    
    // Implementation of the iteration logic
    func serveDinner() {
        let eatingStyle = createDinner()
        for dish in dishes {
            eatingStyle.eat(dish: dish)
        }
    }
}

// --- Main Execution ---

let myWeekendDinner = WeekendDinner()
myWeekendDinner.add(dish: FoodDish(name: "Steak Tartare"))
myWeekendDinner.add(dish: FoodDish(name: "Lobster Thermidor"))
myWeekendDinner.add(dish: FoodDish(name: "Chocolate Soufflé"))

print("Starting Weekend Dinner:")
myWeekendDinner.serveDinner()
